File: c++/Contests/ONI_9/ONI_9/2006 Targoviste/ziua2/solutie/fact/fact.pas
program factorial;
const t=5;
type matrice=array[0..10,0..2] of longint;
var  n:longint;
     b,nr,i:longint;
     d:matrice;
     fin,fout:text;
procedure desc(b:longint);
var i:longint;
begin
     nr:=0;
     i:=2;
     while b>1 do
           begin
                if b mod i=0 then
                   begin
                        d[nr,0]:=i;
                        d[nr,1]:=0;
                        while b mod i=0 do
                              begin
                                   d[nr,1]:=d[nr,1]+1;
                                   b:=b div i
                              end;
                        nr:=nr+1;
                   end;
                i:=i+1;
           end;
end;
function cifra:longint;
var i,aux:longint;
    j,p:longint;
begin
     p:=1;
     for j:=0 to nr-1 do
         d[j,2]:=0;
     for i:=1 to n do
         begin
              aux:=i;
              for j:=0 to nr-1 do
                  while aux mod d[j,0]=0 do
                        begin
                             d[j,2]:=d[j,2]+1;
                             aux:=aux div d[j,0];
                        end;
              p:=(p*(aux mod b))mod b;
         end;
     cifra:=p;
end;
function minim:longint;
var i:longint;
    min:longint;
begin
     min:=d[0,2] div d[0,1];
     for i:=1 to nr-1 do
         if d[i,2] div d[i,1]<min
            then min:=d[i,2] div d[i,1];
     minim:=min;
end;
function calcul:longint;
var i,j,p:longint;
    min:longint;
begin
     p:=cifra;
     min:=minim;
     for i:=0 to nr-1 do
         for j:=1 to d[i,2]-d[i,1]*min do
             p:=(p*(d[i,0] mod b))mod b;
     calcul:=p;
end;
procedure scrie;
var x:longint;
begin
     x:=calcul;
     if x<10
        then writeln(fout,x)
        else writeln(fout,chr(ord('A')+x-10))
end;
begin
     assign(fin,'fact.in');
     reset(fin);
     assign(fout,'fact.out');
     rewrite(fout);
     for i:=1 to 5 do
         begin
              readln(fin,n,b);
              desc(b);
              scrie;
         end;
     close(fin);
     close(fout)
end.

File: c++/Contests/ONI_9/ONI_9/2006 Targoviste/ziua2/solutie/fact/fact.cpp
#include<stdio.h>
#define lmax 10000

FILE *fin,*fout;

void multiply(long x[lmax],long y,int b)
{
 long p[lmax],i,j,t,aux;

 p[0]=x[0];
 for(i=1;i<lmax;i++) p[i]=0;
 t=0;
 for(i=1;i<=x[0];i++)
  {
   p[i]=x[i]*y+t;
   t=p[i]/b;
   p[i]=p[i]%b;
  }
 while(t)
 {
  p[i]=t%b;
  t=t/b;
  i++;
 }
 p[0]=i-1;
 for(i=0;i<=p[0];i++) x[i]=p[i];
}

void print(long x[lmax])
{
 int i;

 i=1;
 while(!x[i]) i++;
 if(x[i]<10) fprintf(fout,"%ld\n",x[i]);
  else fprintf(fout,"%c\n",'A'+x[i]-10);
}

int main()
{
 long n,b,i,j;
 long x[lmax];

 fin=fopen("fact.in","r");
 fout=fopen("fact.out","w");
 for(i=1;i<=5;i++)
 {
  fscanf(fin,"%ld %ld",&n,&b);
  x[0]=x[1]=1;
  for(j=1;j<=n;j++)
   multiply(x,j,b);
  print(x);
  for(j=0;j<lmax;j++) x[j]=0;
 }
 fclose(fin);
 fclose(fout);
 return 0;
}
